### PR TITLE
Apptainer environment override, Python drop/bump, pyhit-wasp recipe marriage, versioner QOL

### LINKED ISSUE #28813: create a 99-end-user.sh override ability in our Apptainer containers
### Motivation

`90-environment.sh`:
```bash
    # Load a custom environment if it exists
    if [ -e /init_env ]; then
        source /init_env
    fi
```
The above is not enough to support bash functions in `/init_env`.

### Design

After some trial and error, we need to go back to using `set -- --noprofile --rcfile`.

Create a method to only do these options if `/init_env` exists. Something like:
```bash
$ cat 99-user-overrides.sh 
#!/bin/sh
# allow user overrides if /init_env exists
# NOTE: `source /init_env` is not enough to support user profile bash functions
if [[ -e /init_env ]]; then
    file='/init_env'
    action="${0##*/}"
    case "${action}" in
    shell)
        if [ "${SINGULARITY_SHELL:-}" = "/bin/bash" ]; then
            set -- --noprofile --rcfile $file
        elif test -z "${SINGULARITY_SHELL:-}"; then
            export SINGULARITY_SHELL=/bin/bash
            set -- --noprofile --rcfile $file
        fi
        ;;
    exec)
        export BASH_ENV="$file"
        set -- /bin/bash --noprofile --rcfile $file -c "$*"
        ;;
    run)
        set -- /bin/bash --noprofile --rcfile $file
    esac
fi
```
### Impact

launching apptainer containers while bind mounting /init_env will support a persons full bash_profile experience. Also, the above example seems to continue to allow all other container functionality (AFAIK).

### PR DESCRIPTION
**Pull request by milljm**
This PR adds the following:

- An `${HOME}/.bashrc_apptainer` user override to be applied as they enter our containers if the file exists.
- Python 3.8 drop (EOL).
- Python 3.12 addition.
- Combines PyHit Conda package into WASP's.
- versioner.py quality of life improvment, by providing a summary of what should be appended to `versioner_hashes.yaml` when bumping an influential file:
   ```bash
   $ ./versioner.py --summary
   8e39fa181b8abba62f117006603d3a1dddea642d: #PR
   tools: 0c9ab93
   mpi: 80648d3
   petsc: 9664f0c
   libmesh: cf37955
   wasp: 98d4d49
   moose-dev: 5410b50
   ```

Closes #28813 #27690

### CODE REVIEW

File: apptainer/files/mpi/.singularity.d/env/99-user-overrides.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+# allow user overrides if /init_env exists
```
Comment (loganharbour) [inline apptainer/files/mpi/.singularity.d/env/99-user-overrides.sh:2]:
```suggestion
# allow user overrides if ${HOME}/.bashrc_apptainer exists
```

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: python/MooseDocs/test/extensions/test_versioner.py [hunk 2]
```diff
@@ -59,8 +59,14 @@ def testCondaVersionRender(self):
...
-            # Application doesn't have a conda version
-            has_conda_version = package != 'app'
+            # Application, doesn't have a conda version
+            # Several support libraries also have no conda version
+            has_conda_version = package not in ['app',
+                                                'libmesh-vtk',
+                                                'peacock',
+                                                'pprof',
+                                                'pyhit',
+                                                'seacas']
```
Comment (milljm) [inline python/MooseDocs/test/extensions/test_versioner.py:69]: @loganharbour I am not familiar with these things not having a 'conda version'. Especially the new tracking additions. But without these, I do get errors. I want to make sure I am doing this right.

error:
```pre
Traceback (most recent call last):
  File "/Users/milljm/projects/moose/python/MooseDocs/test/extensions/test_versioner.py", line 80, in testCodeRender
    conda_version = self.version_meta[package]['conda']['version_and_build']
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
KeyError: 'version_and_build'
```
I suppose I need to see how we're building this dictionary...

File: scripts/tests/versioner_hashes.yaml [hunk 3]
```diff
@@ -355,3 +355,10 @@ b9c32bf26ef2dc29535a2494fb64c7736bcd0d13: #28717
   libmesh: 0e90306
   moose-dev: 28a1964
   wasp: 33b8e6b
+a09522090b8e33988b9566bc3744babbb735b627: #27690 #28813
+  toos: f523cb7
```
Comment (milljm) [inline scripts/tests/versioner_hashes.yaml:359]: lol...

File: conda/functions/retry_build.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            (( try_count++ )) || true
+            printf "Exhausted retry attempts: %s\n" "${try_count}"
+            exit 1
+        fi
+        (( try_count++ )) || true
+        tail -100 output.log
+        printf "\n\nRoutine error caught, trying again...\n"
+        true > "${SRC_DIR}"/output.log
+    done
+}
```
Comment (milljm) [inline conda/functions/retry_build.sh:36]: I am purposely not tracking this file. It has no influence on how packages are built. Only that they get built. I am also not a fan of changes to this file requiring a near-complete entire stack rebuild.

File: conda/seacas/meta.yaml [hunk 5]
```diff
@@ -9,8 +9,9 @@ package:
   version: {{ seacas_version }}
 
 source:
-  git_url: https://github.com/sandialabs/seacas
-  git_rev: {{ seacas_git_rev }}
+  - git_url: https://github.com/sandialabs/seacas
+  - git_rev: {{ seacas_git_rev }}
```
Comment (milljm) [inline conda/seacas/meta.yaml:13]: Maybe this is no longer grabbing the correct versions. Investigate on Monday if latest push still fails.

Comment (milljm) [reply conda/seacas/meta.yaml:13]: Addressed in later commits.

File: apptainer/mpi.def [hunk 6]
```diff
@@ -75,6 +70,9 @@ set_prompt
 EOF
     chmod 755 ${SINGULARITY_ROOTFS}{{ SET_PROMPT_SCRIPT }}
 
+    # insure all profiles are read/executable by everyone
+    chmod +rx ${SINGULARITY_ROOTFS}/.singularity.d/env/*.sh
+
```
Comment (milljm) [inline apptainer/mpi.def:75]: @loganharbour, is this what you were requesting?

File: python/mooseutils/PerfGraphReporterReader.py [hunk 7]
```diff
@@ -9,6 +9,9 @@
 
 from mooseutils.ReporterReader import ReporterReader
 
+# Use old sum() behavior (Python <3.12)
```
Comment (loganharbour) [inline python/mooseutils/PerfGraphReporterReader.py:12]: Is all of the r""" in here a replacement failure?

File: python/mooseutils/ReporterReader.py [hunk 8]
```diff
@@ -14,7 +14,7 @@
 from . import message
 
 class ReporterReader(object):
```
Comment (loganharbour) [inline python/mooseutils/ReporterReader.py:16]: Same here for r"""

**Review by loganharbour — COMMENTED** [review]
Do we want to update python in moose-dev.def to 3.12?

File: python/mooseutils/PerfGraphReporterReader.py (hunk 7 above)
Comment (milljm) [reply python/mooseutils/PerfGraphReporterReader.py:12]: Python 3.12 required it for unittests. If I missed one, I'll add it.

Comment (milljm) [reply python/mooseutils/PerfGraphReporterReader.py:12]: Oh I understand the question.

r""" is required when there are regular expressions within the docstring. Why 3.12 complains about this... 🤷‍♂️ 

That is why some have them, while others do not.

Comment (milljm) [reply python/mooseutils/PerfGraphReporterReader.py:12]: https://discuss.python.org/t/getting-syntax-errors-in-docstrings-how-to-fix/50104

**Comment by milljm** [comment]
> Do we want to update python in moose-dev.def to 3.12?

will do...

File: conda/wasp/meta.yaml [hunk 9]
```diff
@@ -14,25 +13,79 @@ package:
...
-    - {{ pin_subpackage('moose-wasp', max_pin='x.x')}}
 
-requirements:
-  build:
-    - {{ compiler('c') }}
-    - {{ compiler('cxx') }}
-    - make
-    - cmake
-    - pkg-config
+outputs:
```
Comment (loganharbour) [inline conda/wasp/meta.yaml:35]: This looks great. proper pyhit integration

**Review by loganharbour — APPROVED** [review]
Good pending the open conversation above in https://github.com/idaholab/moose/pull/28847#discussion_r1803422517

File: python/MooseDocs/test/extensions/test_versioner.py (hunk 2 above)
Comment (milljm) [reply python/MooseDocs/test/extensions/test_versioner.py:69]: I separated what versioner.py can keep track of, over what unittest is verifying in versioner_hashes.yaml. We don't want to keep track of every single artifact in versioner_hashes.yaml...

File: scripts/versioner.yaml [hunk 10]
```diff
@@ -73,6 +72,29 @@ packages:
...
+  pprof:
+    influential:
+      - conda/pprof
+  pyhit:
+    dependencies:
+      - wasp
+    influential:
+      - conda/wasp/pyhit
+      - conda/wasp/build_pyhit.sh
+      - conda/wasp/pyhit.patch
```
Comment (loganharbour) [inline scripts/versioner.yaml:94]:
```suggestion
    influential:
      - conda/wasp/pyhit
      - conda/wasp/build_pyhit.sh
      - conda/wasp/pyhit.patch
      - framework/contrib/hit
```

**Review by loganharbour — CHANGES_REQUESTED** [review]

**Comment by lindsayad** [comment]
hopefully this can get in soon so we can move on with the petsc update

**Review by loganharbour — APPROVED** [review]

**Comment by loganharbour** [comment]
> hopefully this can get in soon so we can move on with the petsc update

We just finished it up. Aiming to get it in tonight.

**Comment by loganharbour** [comment]
Failures are unrelated.

**Pull request merged**